### ISSUE TITLE
Compass directions in GUI is inconsistent with debug screen and vanilla maps when near modded structures/biomes

### ISSUE THREAD

**Issue by The-Bernician-Lamb** [Serilum/.issue-tracker#102]
## **Information**
Minecraft version: 1.16.1
Forge version: 32.0.98
Environment: Multiplayer and Singleplayer

Mod name: GUI Compass
Mod version: 1.3

## **Description**
Initially, when I first load into a world, the gui displayed when I have a compass is correct, but as I explore the world and generate new chunks it starts to become incorrect and inconsistent with vanilla maps and the debug screen (F3). What mainly tends to happen is that the GUI for this mod gets East and West the wrong way round and leads me west when I am wanting to go East and Vice versa, though North and South tend to always be correct regardless of how far I explore. It only seems to be east and west that the GUI gets confused with.

The GUI also seems to be slightly skewed in relation to vanilla maps, in that when I'm facing a certain direction as displayed on a vanilla map, it displays a completely different direction in the gui or one that is slightly off from where it should be.  (E.g displaying North in the GUI, when the map is telling me I'm facing North-East or North-west and vice versa).

Update: I have found that you can easily replicate this issue by teleporting to a different location via a teleport command, after teleporting the compass GUI does not re-align correctly. This can be done either through vanilla commands or methods added by other mods.

**Comment by ricksouth** [comment]
I can't seem to replicate it in my local world. Are you using singleplayer or multiplayer? Could you send me your modlist?

**Comment by The-Bernician-Lamb** [comment]
I'm mainly using Multiplayer right now.

My mod list is as follows:
Abnormals core - 1.0.4
Aquaculture - 2.0.13
Atmospheric - 1.3.0
AutoRegLib - 1.5-40
Autumnity - 1.3.0
Bamboo Blocks - 1.4.0
Bedspreads - 3.0.0.1
Berry Good - 2.0.0
Better Advancements - 0.1.0.96
Better Burning - 1.3.4
Better Caves - 1.0
Better Drowning - 1.0.3
Biome info - 1.2.5
Biomes o Plenty - 10.0.0.361
Bloomful - 1.4.0
Bookshelf - 5.5.36
Bounty of the land - 0.1.2
Breed Killer Rabbit - 1.1
Buzzier Bees - 1.5.0
Caelus - 2.0-beta2
Campfire Torches - 1.0.2
Cauldron Recipes - 1.1.5
Charm - 1.5.7
Clumps - 5.0.2
cofh Core - 0.4.0b
Colytra - 3.0
Comforts - 2.0.0.2
Controlling - 6.1.4
Cooking For Blockheads - 8.0.1
Cosmetic Armour Reworked - v3a
Deadly End Phantoms - 1.01
Ding - 1.1.0
Ensorcellation - 0.4.1b
Fast Furnace - 3.0.0
Fast Leaf Decay - v22
Fast Workbench - 3.0.0
Guard Villagers - 1.0.6
GUI Compass - 1.1
Hunter Illager - 0.7.0
Hwyla - 1.10.8-B72
Incubation - 1.0.4
JEI - 6.0.0.4
Just Enough Resources - 0.10.1.77
Leaves decay on other leaves - 1.1
Lithium - 0.4.6 - rc3
Login Protection - 1.2
Mantle - 1.5.15
Mixin Bootstrap - 1.0.3
Mooshroom Tweaks - 1.0
Naturally Charged Creepers - 1.1
Oh my Gourd - 0.1.0b
Optiforge - 0.1.18
Phosphor - 0.5.2 + build4
Pick up Notifier - 1.1.2
Placebo - 3.0.2
Polarizing biomes - 1.0
Optifine - 1.15.2 HD U G1 Pre17
Project Vibrant Journeys - 2.0.1
Quark Oddities - 1.15.2
Quark - r2.1-237
Random Patches - 1.21.1.2
Reauth - 3.8.1
Repurposed Structures - 1.4.1
Respawnable Pets - 1.1
Savage and Ravage - 1.1.0
Save my Stronghold - 1.0.0
Simple Farming - 1.2.7
Skeleton Horse Spawn - 1.1.2
Snow under trees - 1.1
Spice of Life: Carrot edition - 1.9.4
Strange - 1.0.0-beta3
Swamp Expansion - 1.7.1
Swing through grass - 1.4.0
Teams mod - 1.7
The Endergetic Expansion - 1.3.1
Tinker's Mechworks - 2.1.0
Toast Control - 3.0.1
Corail Tombstone - 4.3.7
Upgrade Aquatic - 1.6.1
Valhelsia Structures - 15.0.2a
Venture - 1.0.0
Wither Skeleton Tweaks - 4.0.0

**Comment by The-Bernician-Lamb** [comment]
After some testing I have also been able to replicate this issue in a singleplayer world, though only after I had generated a significant amount of chunks. I did this by using the /locate command to locate various structures both vanilla and modded and then teleporting to them via the coordinates that were printed in the chat. (Though this takes significantly less time on a multiplayer server). To give you a better idea of what I mean when I say the gui is inconsistent I have included some screenshots:
When facing North:
![2020-05-17_13 53 07](https://user-images.githubusercontent.com/64381954/82147268-884b7280-9846-11ea-8d0a-b29e06adaa1c.png)
When facing East:
![2020-05-17_13 53 20](https://user-images.githubusercontent.com/64381954/82147320-d6f90c80-9846-11ea-9eab-9dd572282f01.png)
When facing South:
![2020-05-17_13 53 37](https://user-images.githubusercontent.com/64381954/82147347-04de5100-9847-11ea-9f14-524213fd9fe3.png)
When facing West:
![2020-05-17_13 53 50](https://user-images.githubusercontent.com/64381954/82147363-1c1d3e80-9847-11ea-8b09-9cfe831aa086.png)

**Comment by ricksouth** [comment]
Thanks a lot for the detailed information! I'm going to try to fix it. Right now I'm thinking a force-resync every time a player teleports or after a certain amount of time. But there might be an easier solution. I'll keep you updated

**Comment by The-Bernician-Lamb** [comment]
I've been doing a bit more testing around this issue and I found out some things that may be of interest:
1. Exploring long distances (especially underground) seems to be the main cause of this issue.
2. Sleeping in a bed and skipping the night causes the gui to sync up again and go back to showing the correct compass directions. However this does not stop it from de-syncing again after a while.

I don't know whether this is actually any use or not, I just thought it was something interesting that might help you with this issue. 
I appreciate the work that you do, your mods are great! Keep up the good work!

**Comment by ricksouth** [comment]
Again, thanks for the detail! I've uploaded version 1.2 to curseforge which I hope fixes the problem. But I can't seem to replicate the issue on my development pc. Even after long exploring and /locate location teleporting. (Although it might not be long enough O:).

I assume you did all testing with the other mods enabled right? The next step for me might be to install them all as well, if 1.2 doesn't work.

**Comment by The-Bernician-Lamb** [comment]
Thanks! I'll give it a try and get back to you about it!

Yes, I did test with all other mods enabled. So it may be an issue in relation to another mod if this update didn't fix it.

**Comment by The-Bernician-Lamb** [comment]
So, I've done some testing regarding this issue. My results are as follows:

The patch goes a long way to help fix this issue in relation to vanilla biomes and structures and as such no longer occurs when near vanilla structures and biomes. 

The re-sync when teleporting is very helpful and very much helps to re-orient myself after I've teleported to a new location.

However this patch has not gotten rid of the issue completely. I've found the issue still crops up when I'm exploring biomes and structures implemented by other mods such as Biomes o plenty, Project: Vibrant Journeys, Strange and Repurposed structures. When I'm exploring vanilla biomes and structures everything seems to stay synced up properly with little to no issues. 

Whenever I use /locate to teleport to a structure added by another mod, it typically stays synced for a bit due to the re-sync implemented when teleporting, but then de-syncs again after about 30 seconds to a minute of exploring the area.

Overall, This patch goes a long way to fixing this issue but does not eradicate it completely when other mods are involved that change worldgen in some way.

Update: Modded Structures definitely play a much bigger role in creating this issue than modded biomes do. In fact I think it may just be modded structures causing the de-sync. Whenever I teleport to a structure that has been added by a mod (Or a vanilla structure that has been modified by a mod) then it takes no more than 30 seconds for this de-sync to occur. (Definitely keep the fixes already implemented though, they help a lot!)

The mods that seems to be causing this issue the most is Hunter Illager (when I teleport to the Hunter House structure it adds) and Savage and Ravage (When I teleport to the vanilla pillager outpost which it modifies) and Valhelsia structures (Sometimes when I teleport to some of it's surface structures).
